ion-schema/src/constraint.rs

```diff
@@ -1031,6 +1067,117 @@ impl ConstraintValidator for FieldsConstraint {
+}
+
+impl FieldNamesConstraint {
+    pub fn new(type_reference: TypeReference, requires_distinct: bool) -> Self {
+        Self {
+            type_reference,
+            requires_distinct,
+        }
+    }
+}
+
+impl ConstraintValidator for FieldNamesConstraint {
+    fn validate(
+        &self,
+        value: &IonSchemaElement,
+        type_store: &TypeStore,
+        ion_path: &mut IonPath,
+    ) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // create a set for checking duplicate elements
+        let mut element_set = vec![];
+
+        // get elements for given container in the form (ion_path_element, element_value)
+        let elements: Vec<(IonPathElement, &Element)> = match value {
+            IonSchemaElement::SingleElement(element) => {
+                // Check for null struct
+                if element.is_null() {
+                    return Err(Violation::new(
+                        "field_names",
+                        ViolationCode::TypeMismatched,
+                        "expected a struct but found null",
+                        ion_path,
+                    ));
+                }
+
+                return match element.as_struct() {
+                    Some(struct_value) => {
+                        let field_names: Vec<(IonPathElement, &Symbol)> = struct_value
+                            .iter()
+                            .map(|(field_name, val)| {
+                                (
+                                    IonPathElement::Field(field_name.text().unwrap().to_owned()),
+                                    field_name,
+                                )
+                            })
+                            .collect();
+                        // validate field_names constraint
+                        for (ion_path_element, field_name) in field_names {
+                            ion_path.push(ion_path_element);
+                            let schema_element: IonSchemaElement =
+                                (&Element::symbol(field_name)).into();
+
+                            if let Err(violation) =
+                                self.type_reference
+                                    .validate(&schema_element, type_store, ion_path)
+                            {
+                                violations.push(violation);
+                            }
+                            if self.requires_distinct && element_set.contains(&field_name) {
+                                violations.push(Violation::new(
+                                    "field_names",
+                                    ViolationCode::ElementNotDistinct,
+                                    format!("expected distinct field names but found duplicate field name {field_name}", ),
+                                    ion_path,
+                                ))
+                            }
+                            element_set.push(field_name);
```

Comment:
We don't need to add things to `element_set` unless we're checking for duplicates, and we don't need to add a field name to `element_set` if we've already determined that it's a duplicate.

```suggestion
                            if self.requires_distinct {
                                if element_set.contains(&field_name) {
                                    violations.push(Violation::new(
                                        "field_names",
                                        ViolationCode::ElementNotDistinct,
                                        format!("expected distinct field names but found duplicate field name {field_name}", ),
                                        ion_path,
                                    ))
                                } else {
                                    element_set.push(field_name);
                                }
                            }
```
This is just one possible way of dealing with the things I mentioned above. It _might_ be more readable to invert the inner `if` to `if not in set { add to set } else { add violation }`, or even factor it a different way. Your call.